pkg/ebpfsyncer/ebpfsyncer.go

```diff
@@ -80,93 +84,176 @@ func (e *ebpfSingleton) SyncInterfaceIngressRules(
-			_, err := e.c.IngressNodeFwAttach([]string{intf}, true)
-			if err != nil {
-				return err
+// attachFailsafeRules attaches failsafe rules to the provided set of IngressNodeFirewallRules. It modifies the provided
+// map in place.
+func (e *ebpfSingleton) attachFailsafeRules(ifaceIngressRules map[string][]v1alpha1.IngressNodeFirewallRules) error {
+	for _, ingressRules := range ifaceIngressRules {
+		for i := range ingressRules {
+			if err := addFailSaferules(&ingressRules[i].FirewallProtocolRules); err != nil {
+				return fmt.Errorf("Fail to attach fail safe rules, rule: %v, err: %q", ingressRules[i], err)
 			}
 		}
-		// TODO: Clean up all rules that were associated to this interface - somehow part of e.c.'s functionality?
 	}
+	return nil
+}
 
-	// Add rules for both managed and unmanaged interfaces given that the add operation is already idempotent.
-	// TODO missing functionality to purge undesired rules - should be part of e.c's functionality?
-	// Diff both the provided and the currently known IngressNodeFirewallRules interfaces.
-	// Iterate over the current known IngressNodeFirewallRules and purge anything that should not be there.
-	for intf, ingress := range ifaceIngressRules {
-		if _, ok := e.managedInterfaces[intf]; !ok {
-			// Attach to the interfaces.
-			logger.Info("Attaching firewall interface", "intf", intf)
-			_, err := e.c.IngressNodeFwAttach([]string{intf}, isDelete)
-			if err != nil {
-				logger.Error(err, "Fail to attach ingress firewall prog")
-				return err
-			}
-			e.managedInterfaces[intf] = struct{}{}
-		}
+// loadIngressNodeFirewallRules adds, updates and deletes rules from the ruleset.
+func (e *ebpfSingleton) loadIngressNodeFirewallRules(
+	ifaceIngressRules map[string][]v1alpha1.IngressNodeFirewallRules) error {
+	e.log.Info("Loading rules")
+	if err := e.c.IngressNodeFwRulesLoader(ifaceIngressRules); err != nil {
+		e.log.Error(err, "Failed loading ingress firewall rules", ifaceIngressRules, ifaceIngressRules)
+		return err
+	}
+	return nil
+}
 
-		// Look up the network interface by name.
-		iface, err := net.InterfaceByName(intf)
+// resetAll deletes all current attachments and cleans all eBPFObjects. It then sets the ingress firewall manager
+// back to nil. It also deletes all pins and removed all XDP attachments for all system interfaces.
+func (e *ebpfSingleton) resetAll() error {
+	e.log.Info("Running detach operation of managed interfaces")
+	for intf := range e.managedInterfaces {
+		err := e.c.IngressNodeFwDetach(intf)
 		if err != nil {
-			return fmt.Errorf("lookup error for network iface %q: %s", intf, err)
+			e.log.Info("Could not detach managed interface", "intf", intf, "err", err)
 		}
-		ifId := uint32(iface.Index)
+	}
+
+	e.log.Info("Running XDP detach operation from all interfaces")
+	if err := intutil.DetachXDPFromAllInterfaces(); err != nil {
+		e.log.Info("Error during detach operation", "err", err)
+	}
+
+	e.log.Info("Removing all pins")
```

Comment:
why don't we combine all those in one method to export less details to this pkg ? it doesn't need to know anything about pinDir etc ?

Reply:
See my comment aboeve but agreed, I'll move them back into one method but I'll keep them modular as private methods for the future just in case we need to troubleshoot this further

Reply:
done